src/gp_instance.cpp

```diff
@@ -86,3 +130,805 @@ void GPInstance::ProcessLoadedTrees() {
+    RootwardDepthFirst(leaf_idx,
+                       dag_nodes_,
+                       visit_order,
+                       visited_nodes);
+  }
+  
+  return visit_order;
+}
+
+std::vector<size_t> GPInstance::RootwardPassTraversal()
+{
+  std::vector<size_t> visit_order;
+  std::unordered_set<size_t> visited_nodes;
+  for (auto rootsplit : rootsplits_) {
+    size_t root_idx = subsplit_to_index_[rootsplit + ~rootsplit];
+    LeafwardDepthFirst(root_idx,
+                       dag_nodes_,
+                       visit_order,
+                       visited_nodes);
+  }
+  return visit_order;
+}
+
+void GPInstance::BuildPCSPIndexer()
+{
+  size_t taxon_count = alignment_.SequenceCount();
+  size_t idx = 0;
+  for (auto rootsplit : rootsplits_) {
+    SafeInsert(pcsp_indexer_, rootsplit + ~rootsplit, idx);
+    idx++;
+  }
+
+  for (size_t i = taxon_count; i < dag_nodes_.size(); i++) {
+    auto node = dag_nodes_[i];
+    auto n_child = node->GetLeafwardSorted().size();
+    if (n_child > 0) {
+      SafeInsert(subsplit2range_, node->GetBitset(), {idx, idx + n_child});
+      for (size_t j = 0; j < n_child; j++) {
+        auto child = dag_nodes_[node->GetLeafwardSorted()[j]];
+        SafeInsert(pcsp_indexer_, node->GetBitset() + child->GetBitset(), idx);
+        idx++;
+      }
+    }
+    n_child = node->GetLeafwardRotated().size();
+    if (n_child > 0) {
+      SafeInsert(subsplit2range_,
+                 node->GetBitset().RotateSubsplit(),
+                 {idx, idx + n_child});
+      for (size_t j = 0; j < n_child; j++) {
+        auto child = dag_nodes_[node->GetLeafwardRotated()[j]];
+        SafeInsert(pcsp_indexer_,
+                   node->GetBitset().RotateSubsplit() + child->GetBitset(),
+                   idx);
+        idx++;
+      }
+    }
+  }
+}
+
+void GPInstance::InitializeGPEngine()
+{
+  size_t n_params = engine_->GetBranchLengths().size();
+
+  // Initialize branch lengths to 1.
+  EigenVectorXd branch_lengths(n_params);
+  branch_lengths = EigenVectorXd::Ones(n_params);
+  GetEngine()->SetBranchLengths(branch_lengths);
+
+  // For now, use uniform.
+  EigenVectorXd q = EigenVectorXd::Ones(n_params);
+  q.segment(0, rootsplits_.size()).array() = 1./rootsplits_.size();
+  for (auto it = subsplit2range_.begin(); it != subsplit2range_.end(); ++it) {
+    auto range = subsplit2range_[it->first];
+    auto num_child_subsplits = range.second - range.first;
+    double val = 1./num_child_subsplits;
+    q.segment(range.first, num_child_subsplits).array() = val;
+  }
+  GetEngine()->SetSBNParameters(q);
+
+  // Set rhat(s) = stationary for the rootsplits s.
+  std::vector<GPOperation> operations;
+  for (size_t i = 0; i < rootsplits_.size(); i++) {
+    auto rootsplit = rootsplits_[i];
+    auto root_idx = subsplit_to_index_[rootsplit + ~rootsplit];
+    operations.push_back(SetToStationaryDistribution{
+      GetPlvIndex(PlvType::R_HAT, dag_nodes_.size(), root_idx),
```

Comment:
May I request "PLV" to agree with other such abbreviations in the codebase, e.g. "SBN", and even related code such as "PopulatePLVs"?

Reply:
Done.